core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -59,13 +60,16 @@
   private final StoreListener<K, V> storeListener = new StoreListener<K, V>();
   private final Store<K, V> store;
   private final TimeSource timeSource;
+  private final ConcurrentHashMap<K, ConcurrentLinkedQueue<CacheEventWrapper>> keyBasedEventQueueMap = new ConcurrentHashMap<K, ConcurrentLinkedQueue<CacheEventWrapper>>();
```

Comment:
Not an option to have a map with one queue per key. You need to have something that spreads key over different queues indeed but no a 1-1 mapping.